pkg/storage/mysql/mysqlstore.go

```diff
@@ -0,0 +1,271 @@
+/*
+Copyright SecureKey Technologies Inc. All Rights Reserved.
+SPDX-License-Identifier: Apache-2.0
+*/
+
+package mysql
+
+import (
+	"context"
+	"database/sql"
+	"errors"
+	"fmt"
+	"strings"
+	"sync"
+
+	// Add as per the documentation - https://github.com/go-sql-driver/mysql
+	_ "github.com/go-sql-driver/mysql"
+
+	"github.com/hyperledger/aries-framework-go/pkg/storage"
+)
+
+// Provider represents an MySQL DB implementation of the storage.Provider interface
+type Provider struct {
+	dbPath string
+	db     *sql.DB
+	dbs    map[string]*sqlDBStore
+	sync.RWMutex
+}
+
+type sqlDBStore struct {
+	db     *sql.DB
+	dbName string
+}
+
+const (
+	blankDBPathErrMsg         = "DB URL for new mySQL DB provider can't be blank"
+	failToCloseProviderErrMsg = "failed to close provider"
+	//nolint:varcheck deadcode
+	tablePrefix   = "t_"
+	sqlDBNotFound = "no rows"
+)
+
+type result struct {
+	key   string
+	value []byte
+}
+
+// NewProvider instantiates Provider
+func NewProvider(dbPath string) (*Provider, error) {
+	if dbPath == "" {
+		return nil, errors.New(blankDBPathErrMsg)
+	}
+
+	// Example DB Path root:my-secret-pw@tcp(127.0.0.1:3306)/
+	db, err := sql.Open("mysql", dbPath)
+	if err != nil {
+		return nil, fmt.Errorf("failed to open connection: %w", err)
+	}
+
+	p := &Provider{
+		dbPath: dbPath,
+		db:     db,
+		dbs:    map[string]*sqlDBStore{}}
+
+	return p, nil
+}
+
+// OpenStore opens and returns new db for given name space.
+func (p *Provider) OpenStore(name string) (storage.Store, error) {
+	// creating the database
+	_, err := p.db.Exec("CREATE DATABASE IF NOT EXISTS `name`")
+	if err != nil {
+		return nil, fmt.Errorf("failed to create db: %w", err)
+	}
+
+	// Use is used to select the created database without this DDL operations are not permitted
+	_, err = p.db.Exec("USE `name`")
+	if err != nil {
+		return nil, fmt.Errorf("failed to use db: %w", err)
+	}
+
+	store := &sqlDBStore{
+		db:     p.db,
+		dbName: name}
+
+	p.dbs[name] = store
+
+	return store, nil
+}
+
+// Close closes the provider.
+func (p *Provider) Close() error {
+	p.Lock()
+	defer p.Unlock()
+
+	for _, store := range p.dbs {
+		err := store.db.Close()
+		if err != nil {
+			return fmt.Errorf(failToCloseProviderErrMsg+": %w", err)
+		}
+	}
+
+	if err := p.db.Close(); err != nil {
+		return err
+	}
+
+	p.dbs = make(map[string]*sqlDBStore)
+
+	return nil
+}
+
+// CloseStore closes a previously opened store
+func (p *Provider) CloseStore(name string) error {
+	p.Lock()
+	defer p.Unlock()
+
+	store, exists := p.dbs[name]
+	if !exists {
+		return nil
+	}
+
+	delete(p.dbs, name)
+
+	conn, err := store.db.Conn(context.Background())
```

Comment:
should just be closing `store.db` - remember what I mentioned here: https://github.com/hyperledger/aries-framework-go/pull/1938/files#r442128944

Reply:
As per the other store implementations we have two functions close provider which basically closes the store.db and then close store which closes the specific opened connection.

Reply:
updated